### PR TITLE
append serverName to tcp ext_auth destination

### LINKED ISSUE #19374: envoy.filters.network.ext_authz populate unexpected destination with tls_inspector listener
*Title*: envoy.filters.network.ext_authz populate wrong destination with tls_inspector listener

*Description*:
> I am using envoy as a transparent proxy, with ext_auth, seeing envoy send grpc checkRequest with destination to envoy itself, while expect envoy set destination to the original request.

*Repro steps*:
> 1. run envoy container with this configuration file: https://gist.github.com/haitch/e1d6655cffed7a4ee009100d66d30d53#file-envoy_tls_ext_auth-yaml. 
>     docker run --rm -it -p 443:10000 -v envoy-sni.yaml:/etc/envoy.config.yaml envoy-frontend
> 2. have a ext_auth grpc server always return ok, running on localhost:6000
> 3. issue a curl command "curl --resolve www.bing.com:443:127.0.0.1 https://www.bing.com"
> 4. everything works, expect the ext_auth grpc server not getting destination information to inspect the request. (expect destination to be www.bing.com, while actually it is   destination { address { socket_address { address: "172.17.0.2", port_value: 10000 } } }, which is envoy container ip)

*Config*:
> https://gist.github.com/haitch/e1d6655cffed7a4ee009100d66d30d53#file-envoy_tls_ext_auth-yaml

*Logs*:
> https://gist.github.com/haitch/e1d6655cffed7a4ee009100d66d30d53#file-envoy_tls_ext_auth-log

### PR DESCRIPTION
**Pull request by haitch**
Commit Message: append requestServerName to external auth service for better judgement,
 for TCP ext_authz filter (there is another ext_authz for http, which is not touched)

Additional Description: by looking at the protobuf files in /api/envoy/service/auth/v3/attribute_context.proto
.Destination.Service name seems to be perfect property to hold requestServerName
```
message AttributeContext{
  message Peer{
    // The canonical service name of the peer.
    // It should be set to :ref:`the HTTP x-envoy-downstream-service-cluster
    // <config_http_conn_man_headers_downstream-service-cluster>`
    // If a more trusted source of the service name is available through mTLS/secure naming, it
    // should be used.
    string service = 2;
  }
}
```

Risk Level: Low (to my knowledge, this field is left empty to begin with)
Testing: I've compiled and done a manual testing.
Docs Changes:  attached some example request payload for /docs/root/configuration/listeners/network_filters/ext_authz_filter.rst
Release Notes: 
Platform Specific Features:
[Optional Runtime guard:]
fix #19374 
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: test/extensions/filters/common/ext_authz/check_request_utils_test.cc [hunk 1]
```diff
@@ -106,15 +106,17 @@ class CheckRequestUtilsTest : public testing::Test {
   NiceMock<Envoy::StreamInfo::MockStreamInfo> req_info_;
   Buffer::InstancePtr buffer_;
   const std::string cert_data_{"cert-data"};
+  const absl::string_view requestedServerName = "server.name";
```
Comment (dio) [inline test/extensions/filters/common/ext_authz/check_request_utils_test.cc:109]: Nit: based on https://github.com/envoyproxy/envoy/blob/main/STYLE.md

```suggestion
  const absl::string_view requested_server_name_{"server.name"};
```

File: docs/root/configuration/listeners/network_filters/ext_authz_filter.rst [hunk 2]
```diff
@@ -58,6 +58,30 @@ A sample filter configuration could be:
                   address: 127.0.0.1
                   port_value: 10003
 
+A sample request to the specific auth service looks like
```
Comment (dio) [inline docs/root/configuration/listeners/network_filters/ext_authz_filter.rst:61]: Could you expand this, probably telling that this is the body (or context?).

**Review by dio — CHANGES_REQUESTED** [review]
Thanks for working on this @haitch! 

Please add this to https://github.com/envoyproxy/envoy/blob/main/docs/root/version_history/current.rst. Since we want to let users know about this change (you can check on https://github.com/envoyproxy/envoy/blob/main/docs/root/version_history/v1.21.0.rst as a reference).

A couple of nits:

File: docs/root/version_history/current.rst [hunk 3]
```diff
@@ -9,6 +9,7 @@ Incompatible Behavior Changes
 Minor Behavior Changes
 ----------------------
 *Changes that may cause incompatibilities for some users, but should not for most*
+* network.ext_authz filter append destination hostname for authorization service to review.
```
Comment (dio) [inline docs/root/version_history/current.rst:12]: Usually, we do something like: 

```
* ext_authz: added requested server name ...
```

Comment (haitch) [reply docs/root/version_history/current.rst:12]: much simpler, still want to keep "network." prefix, since we have http.ext_authz as well.

Comment (dio) [reply docs/root/version_history/current.rst:12]: Yeah sorry for this nit. You can add that in the sentence after `:`, e.g. https://github.com/envoyproxy/envoy/blob/main/docs/root/version_history/v1.21.0.rst#bug-fixes see the ext_authz entry there.

**Comment by haitch** [comment]
@dio @gsagula Any new comments? Can I get this merged?

**Review by dio — APPROVED** [review]
Thanks!

**Comment by dio** [comment]
/assign-from @envoyproxy/senior-maintainers 

for final pass and merging. Thanks!

**Review by alyssawilk — APPROVED** [review]
code LGTM.  I'll defer to @dio for final merge because my one question is if this should be runtime guarded.

**Comment by alyssawilk** [comment]
@dio thoughts?

**Comment by haitch** [comment]
thanks @alyssawilk  @dio, I solved a merge conflict and cleared all the approve, can you approve again?

also I think we don't need a runtime guard, since it filling a property that was not used. unless someone is doing "==" on the destination, which is very unlikely, since people tend to do subnet/CIDR calculation instead of doing "==".

but, I will leave the decision to @dio .

**Review by alyssawilk — APPROVED** [review]

**Comment by alyssawilk** [comment]
Given Dio's lagging I'll buy that and merge :-)

**Pull request merged**